torch/distributed/rpc/__init__.py

```diff
@@ -71,7 +76,38 @@ def init_rpc(
                 are available.
         """
 
-        if not rpc_backend_options:
+        # To avoid breaking users that passed a ProcessGroupRpcBackendOptions
+        # without specifying the backend as PROCESS_GROUP when that was the
+        # default, we try to detect the backend from the options when only the
+        # latter is passed.
+        if backend is None and rpc_backend_options is not None:
+            for candidate_backend in BackendType:
+                if isinstance(
+                    rpc_backend_options,
+                    type(
+                        backend_registry.construct_rpc_backend_options(
+                            candidate_backend
+                        )
+                    ),
+                ):
+                    backend = candidate_backend
+                    break
+            else:
+                raise TypeError(
+                    f"Could not infer backend for options {rpc_backend_options}"
+                )
+            if backend != BackendType.TENSORPIPE:
+                logger.warning(
+                    f"RPC was initialized with no explicit backend but with options "
+                    f"corresponding to {backend}, hence that backend will be used "
+                    f"instead of the default {BackendType.TENSORPIPE}. To silence this "
+                    f"warning pass `backend={backend}` explicitly."
+                )
+
+        if backend is None:
+            backend = BackendType.TENSORPIPE
+
+        if rpc_backend_options is None:
```

Comment:
What happens if backend is specified as TensorPipe but backend options are ProcessGroupOptions? Do we raise an appropriate error? If so, can we have a unit test that verifies this?